### PR TITLE
Fix: Improve profiler

### PR DESCRIPTION
**Pull request by bout3fiddy**
This PR brings in several improvements to the profiler:

1. Cosmetic changes: the printed out line profiles are a lot more readable
2. remove duplicate entries: previous version had several duplicates
3. Get correct fn_name for line profiles in the case of child computations.
4. introduce `--profile` flag for pytest: this automatically enables profiling for all tests (so you dont need to make a pytest marker for each test).
5. Introduce `ignore_profiling` marker (just in case someone needs it.

image

### CODE REVIEW

File: boa/profiling.py [hunk 1]
```diff
@@ -258,29 +265,33 @@ def cache_gas_used_for_computation(contract, computation):
...
 
     # recursion for child computations
     for _computation in computation.children:
         child_contract = env.lookup_contract(_computation.msg.code_address)
+
         # ignore black box contracts
         if child_contract is not None:
-            cache_gas_used_for_computation(child_contract, _computation)
+            # ignore line profiling for child calls (because it is covered)
+            cache_gas_used_for_computation(child_contract, _computation, True)
```
Comment (charles-cooper) [inline boa/profiling.py:284]: style: please name the last kwarg

File: boa/test/plugin.py [hunk 2]
```diff
@@ -28,17 +28,42 @@ def f(*args, **kwargs):
...
 
+def pytest_collection_modifyitems(config, items):
+
+    if config.getoption("profile"):
+
+        for item in items:
+
+            ignore_isolation = item.get_closest_marker("ignore_isolation")
+            ignore_profiling = item.get_closest_marker("ignore_profiling")
+            if ignore_isolation or ignore_profiling:
```
Comment (charles-cooper) [inline boa/test/plugin.py:51]: why does ignore_isolation turn off profiling?

File: boa/vyper/ast_utils.py [hunk 3]
```diff
@@ -50,3 +50,18 @@ def ast_map_of(ast_node):
     for node in nodes:
         ast_map[getpos(node)] = node
     return ast_map
+
+
+def get_fn_name_from_lineno(ast_map, lineno: int):
+    def _walk_to_fn_node(_node):
+        if _node is None:
+            return ""
+        if _node.ast_type == "FunctionDef":
```
Comment (charles-cooper) [inline boa/vyper/ast_utils.py:59]: use an isinstance here please

File: boa/vyper/ast_utils.py [hunk 4]
```diff
@@ -50,3 +50,18 @@ def ast_map_of(ast_node):
...
+            return _node.name
+        return _walk_to_fn_node(_node.get_ancestor())
+
+    for source_map, node in ast_map.items():
```
Comment (charles-cooper) [inline boa/vyper/ast_utils.py:63]: i don't like the performance of this way of doing things

File: boa/profiling.py [hunk 5]
```diff
@@ -375,10 +411,16 @@ def get_line_profile_table(env: Env) -> Table:
 
                 cname = ""
                 if c == 0:
-                    cname = contract_name + f".{fn_name}"
+                    if len(fn_name) > 30:
+                        fn_name = fn_name[:30] + "..."
```
Comment (charles-cooper) [inline boa/profiling.py:415]: will be cleaner to have a utility like `truncate()` for all of these

File: boa/vyper/ast_utils.py (hunk 4 above)
Comment (bout3fiddy) [reply boa/vyper/ast_utils.py:57]: It worked just about fine for contracts as big as tricrypto (2k+ lines of code).

File: boa/vyper/ast_utils.py [hunk 6]
```diff
@@ -50,3 +51,14 @@ def ast_map_of(ast_node):
     for node in nodes:
         ast_map[getpos(node)] = node
     return ast_map
+
+
+def get_fn_name_from_lineno(ast_map, lineno: int):
+    for source_map, node in ast_map.items():
```
Comment (charles-cooper) [inline boa/vyper/ast_utils.py:57]:
```suggestion
    # TODO: this could be a performance bottleneck
    for source_map, node in ast_map.items():
```

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**